nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -97,7 +97,11 @@ impl From<InstanceAndActiveVmm> for omicron_common::api::external::Instance {
             project_id: value.instance.project_id,
             ncpus: value.instance.ncpus.into(),
             memory: value.instance.memory.into(),
-            hostname: value.instance.hostname,
+            hostname: value
+                .instance
+                .hostname
+                .parse()
+                .expect("found invalid hostname in the database"),
```

Comment:
What if there are existing hostnames in the DB that went in prior to this validation?

Reply:
It's a good question. The change that revealed the [deadlock in the sled-agent](https://github.com/oxidecomputer/omicron/issues/4931) was added in Oct 2023. I am not positive, but it seems unlikely that there are any such instances still in existence.

If we're concerned about it, or want to be more cautious (reasonable!), I can change this `From` implementation to be a `TryFrom` and bubble up a 500 rather than panicking Nexus.

Reply:
> If we're concerned about it, or want to be more cautious (reasonable!), I can change this From implementation to be a TryFrom and bubble up a 500 rather than panicking Nexus.

I think this is the right move - isn't it guaranteed that any instances that had a non-compliant hostname would be in state Failed? We should make sure that those can be deleted, even with the `TryFrom` bubbling up the 500.

Reply:
>  isn't it guaranteed that any instances that had a non-compliant hostname would be in state Failed?

I think that's right, for instances created after Oct 2023. The sled-agent would have deadlocked, and the request from Nexus to it timed out. The result would be a failed instance.

I'll change this now, thanks.

Reply:
This (rightfully) opened a pretty big can of worms for me. The general question is: what do we do with existing invalid hostnames?

The first question I think we need to understand is whether we _have_ invalid hostnames. If we do not, then writing a large amount of code to handle it is not useful.

If we do, there are several ways we can handle it:

- Some kind of data migration during the upgrade. The big pro here is very little new Rust code in the control plane itself. The big con is a lot of manual work, some of which may not even be possible in SQL. For example, I'm not sure how to even detect invalid hostnames in SQL, since the regex engine needs to support lookarounds to work correctly. If it doesn't, we'd need to run more complicated SQL, or potentially run some Rust code directly. We also need to figure out what to do with invalid hostnames.
- Detect and automatically rewrite invalid hostnames when reading them from the database. For example, we could try to validate the hostname every time we read the record, and if that _fails_ we could simply use the instance's `name` field, which is a valid hostname. The pro here is that it's pretty simple to write. The con is obvious: it's confusing and janky, since the instance itself would get a hostname from DHCP that doesn't match the one it was created with.
- The most heavy-weight but careful solution would be to handle both valid and invalid hostnames for some migration period. On writing new instance records, we only insert known-valid hostnames. On read, we would extract the hostname into an enum, something like:

    ```rust
    enum MaybeHostname {
        Valid(Hostname),
        Invalid(String),
    }
    ```

    If we're doing something like fetching or listing the instance for the public API, we could serialize this into a less-rigid view type, which retains the hostname as an unvalidated `String`. If we're doing basically any other operation, most saliently, sending the instance to the sled-agent for it to start, we first validate the hostname.

    The pro here is safety and robustness, at the obvious expense of a large amount of code. We'd need to carry that for some time, but presumably we could remove it again when we're confident there are no extant instances with invalid hostnames.

Reply:
Are we sure that the hostname validation being done by the sled agent is as strict as what you're adding in this PR?

I think it's okay if "no", but just throwing it out there, if your PR rejects a name like `my_host`, and the sled agent would accept it, then the likelihood of invalid hostnames being used seems a lot higher.

My two cents: I like your "Detect and automatically rewrite invalid hostnames" approach, with one small tweak: we could just throw an error, rather than re-writing anything.

For example:

If someone created an instance with an invalid hostname, we could parse the hostname in Nexus during "instance start", and throw an error before sending anything to the sled. Something like "Hostname validation has become more strict, and this name is no longer valid. Please update the hostname or create a new instance".

IMO the burden here isn't terrible -- after an update, we require all instances to stop anyway, so we can still get away with migrations like this. That basically leaves the hostnames alone, but prevents the "bad ones in the field" from being used.

Reply:
> Are we sure that the hostname validation being done by the sled agent is as strict as what you're adding in this PR?
>
> I think it's okay if "no", but just throwing it out there, if your PR rejects a name like my_host, and the sled agent      would accept it, then the likelihood of invalid hostnames being used seems a lot higher.

OPTE's hostname validation is less strict than this PR's. 


> My two cents: I like your "Detect and automatically rewrite invalid hostnames" approach, with one small tweak: we could just throw an error, rather than re-writing anything.

I agree this is a practical approach. It seems like we might be able to get away with checking this in just a few places, such as `Nexus::instance_start()` and maybe `Nexus::instance_reboot()`.

Reply:
So I think I'm going to take this approach, in the interest of urgency. One issue is that the check cannot be when we read instance records from the database, as that would prevent ever operating on those instances at all, including to delete them. Instead, we have to:

- Check hostnames when we start an instance. Luckily, I think there is only one path here, `Nexus::instance_ensure_registered()`, which is called from both the instance start and migration sagas.
- Deserialize hostnames as strings, without validation, from the DB, and send them as strings out the Nexus API. That's to support listing instances and ideally deleting them, even if they have invalid hostnames. We would need to change the instance view type to use a validated hostname in the future, if it matters.

Reply:
Fixed in [4546ed3](https://github.com/oxidecomputer/omicron/pull/4938/commits/4546ed3b7af7131ed0fef2d206e7ee511538bc9d), see testing notes below.